Upstream inline text predictions support for WebKit

rr-codes opened the pull request:
#### 0c989ac562785c6fb59dd5698f926822101141a7
<pre>
Upstream inline text predictions support for WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=257727">https://bugs.webkit.org/show_bug.cgi?id=257727</a>
rdar://110220145

Reviewed by Aditya Keerthi.

Upstream support for system inline text predictions in WebKit.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/page/EditorClient.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::willBeginScrolling):
(WebCore::Page::didFinishScrolling):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration init]):
(-[WKWebViewConfiguration copyWithZone:]):
(-[WKWebViewConfiguration _setMarkedTextInputEnabled:]):
(-[WKWebViewConfiguration _markedTextInputEnabled]):
(-[WKWebViewConfiguration setAllowsInlinePredictions:]):
(-[WKWebViewConfiguration allowsInlinePredictions]):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setCaretDecorationVisibility): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(shouldUseHighlightsForMarkedText):
(extractUnderlines):
(-[WKContentView setAttributedMarkedText:selectedRange:]):
(-[WKContentView setMarkedText:selectedRange:]):
(-[WKContentView _updateTextInputTraits:]):
(-[WKContentView _updateAdditionalTextInputTraits:]): Deleted.
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::setCaretDecorationVisibility): Deleted.
* Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
(WebKit::ViewGestureController::handleMagnificationGestureEvent):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::getTextCheckingTypes const):
(WebKit::WebViewImpl::handleRequestedCandidates):
(WebKit::compositionUnderlines):
(WebKit::WebViewImpl::setMarkedText):
(WebKit::WebViewImpl::allowsInlinePredictions const):
(WebKit::WebViewImpl::showInlinePredictionsForCandidate):
(WebKit::WebViewImpl::showInlinePredictionsForCandidates):
(WebKit::WebViewImpl::shouldRequestCandidates const):
(WebKit::extractInitialUnderlines): Deleted.
(WebKit::WebViewImpl::markedTextInputEnabled const): Deleted.
(WebKit::WebViewImpl::showMarkedTextForCandidate): Deleted.
(WebKit::WebViewImpl::showMarkedTextForCandidates): Deleted.
(WebKit::subscribeToTextInputNotifications): Deleted.
(WebKit::extractUnderlines): Deleted.
(WebKit::WebViewImpl::setCaretDecorationVisibility): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebEditorClientMac.mm:
(WebKit::WebEditorClient::setCaretDecorationVisibility): Deleted.
* Source/WebKitLegacy/mac/WebCoreSupport/WebEditorClient.h:

Canonical link: <a href="https://commits.webkit.org/264903@main">https://commits.webkit.org/264903@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f91a42194241cfb48c61448474c2dea88fbf1cf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9031 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9318 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9537 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10684 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8992 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9039 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11306 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9286 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11843 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9177 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10151 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7910 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10843 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7446 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8251 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15738 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/7769 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8555 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8399 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11728 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/8667 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8901 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7258 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/9198 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8146 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2201 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12361 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/9433 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1046 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8652 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2319 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/page/EditorClient.h:
```diff
@@ -155,8 +155,6 @@ class EditorClient : public CanMakeWeakPtr<EditorClient> {
     virtual void lowercaseWord() = 0;
     virtual void capitalizeWord() = 0;
-
-    virtual void setCaretDecorationVisibility(bool) = 0;
```
This was never actually needed, so I just removed it

aprotyas replied:
We don't need to make this call?

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
```diff
@@ -1492,14 +1488,25 @@ - (_WKContentSecurityPolicyModeForExtension)_contentSecurityPolicyModeForExtensi
     return WebKit::toWKContentSecurityPolicyModeForExtension(_pageConfiguration->contentSecurityPolicyModeForExtension());
 }
 
+// Needed to maintain backwards compatability
```
This should be replaced with a `FIXME` to remove the SPI. Please file bugs against all SPI clients too.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -6255,7 +6319,10 @@ - (void)_updateTextInputTraits:(id<UITextInputTraits>)traits
 
-    [self _updateAdditionalTextInputTraits:privateTraits];
+#if HAVE(INLINE_PREDICTIONS)
+    if ([privateTraits respondsToSelector:@selector(setInlineCompletionType:)])
```
I don't think we need `respondsToSelector` anymore?

Also, if `inlineCompletionType` exists as API, can we just use `traits` rather than `privateTraits`? If not, I think we need something in `UIKitSPI.h`.

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -3152,40 +3152,16 @@ static String commandNameForSelector(SEL selector)
+    NSTextCheckingTypes types = NSTextCheckingTypeSpelling | NSTextCheckingTypeReplacement | NSTextCheckingTypeCorrection;
+#if HAVE(INLINE_PREDICTIONS)
+    if (allowsInlinePredictions())
+        types |= (NSTextCheckingType)_NSTextCheckingTypeSingleCompletion;
```
Is there an API value for `_NSTextCheckingTypeSingleCompletion`? If not, this needs an SPI declaration.

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -5111,6 +5124,113 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
+    if (absoluteSelectedRange.location < relativeSelectedRange.location || absoluteSelectedRange.length != relativeSelectedRange.length)
+        return;
+
+    // Make sure the selected range didn’t change while we were making an asynchronous call to get it
```
Need a period at the end of this sentence.

pxlcoder commented on Source/WebCore/PAL/pal/spi/mac/NSSpellCheckerSPI.h:
```diff
@@ -54,4 +53,6 @@ extern NSString *NSTextCheckingSuppressInitialCapitalizationKey;
 @end
 #endif
 
+#define _NSTextCheckingTypeSingleCompletion (1ULL << 29)
```
We should import the private header underneath `USE(APPLE_INTERNAL_SDK)` and redeclare the enum under `#else`.

pxlcoder commented on Source/WebCore/PAL/pal/spi/mac/NSSpellCheckerSPI.h:
```diff
@@ -45,6 +44,8 @@ extern NSString *NSTextCheckingSuppressInitialCapitalizationKey;
 
 @end
 
+#define _NSTextCheckingTypeSingleCompletion (1ULL << 29)
```
Value should be redeclared matching the private header, `#define` is only for staging changes.

So:

```
typedef NS_OPTIONS(uint64_t, NSTextCheckingTypeAppKitTemporary) {
    _NSTextCheckingTypeSingleCompletion = 1ULL << 29,
};
```

rr-codes replied:
ah, my bad, sorry!

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -5111,6 +5127,113 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
+
+    NSSpellChecker *spellChecker = [NSSpellChecker sharedSpellChecker];
+
+    if (![spellChecker respondsToSelector:@selector(showCompletionForCandidate:selectedRange:offset:inString:rect:view:completionHandler:)])
```
Is this selector check necessary?

I think we also need a declaration in `NSSpellCheckerSPI.h` for OpenSource builds.

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -5111,6 +5127,113 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
+
+    NSSpellChecker *spellChecker = [NSSpellChecker sharedSpellChecker];
+
+    if (![spellChecker respondsToSelector:@selector(completionCandidateFromCandidates:)])
```
Is this selector check necessary?

I think we also need a declaration in `NSSpellCheckerSPI.h` for OpenSource builds.

pxlcoder commented on Source/WebCore/PAL/pal/spi/mac/NSSpellCheckerSPI.h:
```diff
@@ -45,6 +44,10 @@ extern NSString *NSTextCheckingSuppressInitialCapitalizationKey;
 
 @end
 
+typedef NS_OPTIONS(uint64_t, NSTextCheckingTypeAppKitTemporary) {
```
I think this should be behind `HAVE(INLINE_PREDICTIONS)`.

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -5069,6 +5023,68 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
+            underlines.append({ static_cast<unsigned>(range.location), static_cast<unsigned>(NSMaxRange(range)), WebCore::CompositionUnderlineColor::GivenColor, WebCore::Color::black, style.intValue > 1 });
+    }];
+
+    std::sort(underlines.begin(), underlines.end(), [](auto& a, auto& b) {
```
I see similarities between this code and `WKContentViewInteraction.mm`, we should probably refactor at some point (not in this PR).

rr-codes replied:
oh ya definitely; that was my plan, I just forgot to mention it 😅

rr-codes committed:
Upstream inline text predictions support for WebKit
https://bugs.webkit.org/show_bug.cgi?id=257727
rdar://110220145

Reviewed by Aditya Keerthi.

Upstream support for system inline text predictions in WebKit.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/page/EditorClient.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::willBeginScrolling):
(WebCore::Page::didFinishScrolling):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration init]):
(-[WKWebViewConfiguration copyWithZone:]):
(-[WKWebViewConfiguration _setMarkedTextInputEnabled:]):
(-[WKWebViewConfiguration _markedTextInputEnabled]):
(-[WKWebViewConfiguration setAllowsInlinePredictions:]):
(-[WKWebViewConfiguration allowsInlinePredictions]):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setCaretDecorationVisibility): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(shouldUseHighlightsForMarkedText):
(extractUnderlines):
(-[WKContentView setAttributedMarkedText:selectedRange:]):
(-[WKContentView setMarkedText:selectedRange:]):
(-[WKContentView _updateTextInputTraits:]):
(-[WKContentView _updateAdditionalTextInputTraits:]): Deleted.
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::setCaretDecorationVisibility): Deleted.
* Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
(WebKit::ViewGestureController::handleMagnificationGestureEvent):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::getTextCheckingTypes const):
(WebKit::WebViewImpl::handleRequestedCandidates):
(WebKit::compositionUnderlines):
(WebKit::WebViewImpl::setMarkedText):
(WebKit::WebViewImpl::allowsInlinePredictions const):
(WebKit::WebViewImpl::showInlinePredictionsForCandidate):
(WebKit::WebViewImpl::showInlinePredictionsForCandidates):
(WebKit::WebViewImpl::shouldRequestCandidates const):
(WebKit::extractInitialUnderlines): Deleted.
(WebKit::WebViewImpl::markedTextInputEnabled const): Deleted.
(WebKit::WebViewImpl::showMarkedTextForCandidate): Deleted.
(WebKit::WebViewImpl::showMarkedTextForCandidates): Deleted.
(WebKit::subscribeToTextInputNotifications): Deleted.
(WebKit::extractUnderlines): Deleted.
(WebKit::WebViewImpl::setCaretDecorationVisibility): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebEditorClientMac.mm:
(WebKit::WebEditorClient::setCaretDecorationVisibility): Deleted.
* Source/WebKitLegacy/mac/WebCoreSupport/WebEditorClient.h:

Canonical link: https://commits.webkit.org/264903@main

ddkilzer commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.h:
```diff
@@ -143,9 +143,12 @@ WK_CLASS_AVAILABLE(macos(10.10), ios(8.0))
+are disabled regardless of the system setting. If true, they are enabled based
+on the system setting.
+*/
+@property (nonatomic) BOOL allowsInlinePredictions API_AVAILABLE(macos(14.0), ios(17.0));
```
Hi @rr-codes, should this be `WK_API_AVAILABLE()` instead of just `API_AVAILABLE()`?
